qiskit/algorithms/gradients/spsa_estimator_gradient.py

```diff
@@ -0,0 +1,122 @@
+                run_options in ``run`` method > gradient's default run_options > primitive's default
+                setting. Higher priority setting overrides lower priority setting.
+
+        Raises:
+            ValueError: If ``epsilon`` is not positive.
+        """
+        if epsilon <= 0:
+            raise ValueError(f"epsilon ({epsilon}) should be positive.")
+        self._epsilon = epsilon
+        self._batch_size = batch_size
+        self._seed = np.random.default_rng(seed) if seed else np.random.default_rng()
+
+        super().__init__(estimator, **run_options)
+
+    def _run(
+        self,
+        circuits: Sequence[QuantumCircuit],
+        observables: Sequence[BaseOperator | PauliSumOp],
+        parameter_values: Sequence[Sequence[float]],
+        parameters: Sequence[Sequence[Parameter] | None],
+        **run_options,
+    ) -> EstimatorGradientResult:
+        """Compute the estimator gradients on the given circuits."""
+        jobs, offsets, metadata_ = [], [], []
+        for circuit, observable, parameter_values_, parameters_ in zip(
+            circuits, observables, parameter_values, parameters
+        ):
+            # indices of parameters to be differentiated
+            if parameters_ is None:
+                indices = list(range(circuit.num_parameters))
+            else:
+                indices = [circuit.parameters.data.index(p) for p in parameters_]
+            metadata_.append({"parameters": [circuit.parameters[idx] for idx in indices]})
+
+            offset = [
+                (-1) ** (self._seed.integers(0, 2, len(circuit.parameters)))
+                for _ in range(self._batch_size)
+            ]
+
+            plus = [parameter_values_ + self._epsilon * offset_ for offset_ in offset]
+            minus = [parameter_values_ - self._epsilon * offset_ for offset_ in offset]
+            offsets.append(offset)
+
+            job = self._estimator.run(
+                [circuit] * 2 * self._batch_size,
+                [observable] * 2 * self._batch_size,
+                plus + minus,
+                **run_options,
+            )
+            jobs.append(job)
+
+        # combine the results
+        if any(job.status() is not JobStatus.DONE for job in jobs):
+            raise QiskitError("The gradient job was not completed successfully. ")
+        results = [job.result() for job in jobs]
+        gradients = []
+        for i, result in enumerate(results):
+            n = len(result.values) // 2  # is always a multiple of 2
+            gradient_ = (result.values[:n] - result.values[n:]) / (2 * self._epsilon)
+            # take the average of the spsa gradients
+            gradient = np.average(
+                np.array([grad / offset for grad, offset in zip(gradient_, offsets[i])]), axis=0
+            )
+            indices = [circuits[i].parameters.data.index(p) for p in metadata_[i]["parameters"]]
+            gradients.append(gradient[indices])
```

Comment:
I'm not quite sure this is correct: I think you have to evaluate each gradient in the batch, and then average. So something like 
```suggestion
            # compute each gradient in the batch
            n = self._batch_size
            batch_gradients = [
                (result.values[i] - result.values[i + n]) / (2 * self._epsilon * offsets[i])
            ]
            # take the average of the spsa gradients
            gradient = np.mean(batch_gradients, axis=0)

            indices = [circuits[i].parameters.data.index(p) for p in metadata_[i]["parameters"]]
            gradients.append(gradient[indices])
```
Otherwise it seems like you're averaging all plus/minus values and then broadcast it onto the different offsets, which is not quite the same 🙂